sim-lib/src/lib.rs

```diff
@@ -904,48 +905,46 @@ async fn write_payment_results(
     logger: Arc<Mutex<PaymentResultLogger>>,
     mut receiver: Receiver<(Payment, PaymentResult)>,
     listener: Listener,
-    print_batch_size: u32,
-    no_results: bool,
+    write_results: Option<WriteResults>,
 ) -> Result<(), SimulationError> {
-    let mut writer = if !no_results {
-        Some(WriterBuilder::new().from_path(format!(
+    let mut writer = write_results.and_then(|write_result| {
+        let file = write_result.results_dir.join(format!(
             "simulation_{:?}.csv",
             SystemTime::now()
                 .duration_since(SystemTime::UNIX_EPOCH)
                 .unwrap()
                 .as_secs()
-        ))?)
-    } else {
-        None
-    };
+        ));
+        let writer = WriterBuilder::new().from_path(file).ok()?;
+        Some((writer, write_result.batch_size))
+    });
 
     let mut counter = 1;
     loop {
         tokio::select! {
             biased;
             _ = listener.clone() => {
                 log::debug!("Simulation results consumer received shutdown signal.");
-                return writer.map_or(Ok(()), |ref mut w| w.flush().map_err(|_| SimulationError::FileError))
+                return writer.map_or(Ok(()), |(ref mut w, _)| w.flush().map_err(|_| SimulationError::FileError))
             },
             payment_report = receiver.recv() => {
                 match payment_report {
                     Some((details, result)) => {
                         logger.lock().await.report_result(&details, &result);
                         log::trace!("Resolved dispatched payment: {} with: {}.", details, result);
 
-                        if let Some(ref mut w) = writer {
+                        if let Some((ref mut w, batch_size)) = writer {
                             w.serialize((details, result)).map_err(|e| {
                                 let _ = w.flush();
                                 SimulationError::CsvError(e)
                             })?;
-
-                            counter = counter % print_batch_size + 1;
-                            if print_batch_size == counter {
+                            counter = counter % batch_size + 1;
+                            if batch_size == counter {
                                 w.flush().map_err(|_| SimulationError::FileError)?;
                             }
                         }
                     },
-                    None => return writer.map(|ref mut w| w.flush().map_err(|_| SimulationError::FileError)).unwrap_or(Ok(())),
+                    None => return writer.map(|(ref mut w, _)| w.flush().map_err(|_| SimulationError::FileError)).unwrap_or(Ok(())),
```

Comment:
nit: We can use `map_or` instead of `map`  + `unwrap_or` for consistency with the biased branch